core/node/rpc/forwarder.go

```diff
@@ -198,34 +198,79 @@ func (s *Service) getStreamImpl(
 		return nil, err
 	}
 
-	// if the user passed a sync cookie, we need to forward the request to the node that issued the cookie
-	if req.Msg.SyncCookie != nil {
-		nodeAddress := common.BytesToAddress(req.Msg.SyncCookie.GetNodeAddress())
-		if nodeAddress == s.wallet.Address {
-			if view != nil {
-				return s.localGetStream(ctx, view, req.Msg.SyncCookie)
-			} else {
-				return nil, RiverError(Err_BAD_SYNC_COOKIE, "Stream not found").
-					Func("service.getStreamImpl").
-					Tag("streamId", req.Msg.StreamId)
+	cookie := req.Msg.GetSyncCookie()
+
+	// if the user passed a sync cookie, we need to forward the request to the node that issued the cookie.
+	// if the node address is not specified, the sticky peer is used.
+	if cookie != nil {
+		getNodeAddress := func() (common.Address, *SyncCookie, bool) {
+			return common.BytesToAddress(cookie.GetNodeAddress()), cookie, false
+		}
+		if len(cookie.GetNodeAddress()) == 0 {
+			i := 0
+			remotes, useLocal := stream.GetRemotesAndIsLocal()
+			nodeAddress := stream.GetStickyPeer()
+			getNodeAddress = func() (common.Address, *SyncCookie, bool) {
+				if i == 0 && useLocal {
+					useLocal = false
+					return s.wallet.Address, cookie.CopyWithAddr(s.wallet.Address), len(remotes) > 0
+				}
+				if i > 0 {
+					nodeAddress = stream.AdvanceStickyPeer(nodeAddress)
+				}
+				i++
+				return nodeAddress, cookie.CopyWithAddr(nodeAddress), i < len(remotes)
 			}
-		} else {
+		}
+
+		getStream := func(nodeAddress common.Address, cookie *SyncCookie) (*connect.Response[GetStreamResponse], error) {
+			if nodeAddress == s.wallet.Address {
+				if view == nil {
+					return nil, RiverError(Err_BAD_SYNC_COOKIE, "Stream not found").
+						Func("service.getStreamImpl").
+						Tag("streamId", req.Msg.StreamId)
+				}
+
+				return s.localGetStream(ctx, view, cookie)
+			}
+
 			stub, err := s.nodeRegistry.GetStreamServiceClientForAddress(nodeAddress)
-			if err == nil {
-				ret, err := stub.GetStream(ctx, req)
-				if err != nil {
-					return nil, err
+			if err != nil {
+				return nil, err
+			}
+
+			ret, err := stub.GetStream(ctx, req)
```

Comment:
Maybe we should consider passing in a context with a timeout. If one of the nodes is misbehaving and not returning a response within reasonable time we should try another node or in case of non replicated scenario leave it up to the client to retry it.